XenAdmin/Wizards/RollingUpgradeWizard/PlanActions/RpuUploadAndApplySuppPackPlanAction.cs

```diff
@@ -0,0 +1,267 @@
+            }
+        }
+
+        private void PrecheckSuppPack(Pool_update update, string suppPack, out bool alreadyApplied)
+        {
+            alreadyApplied = false;
+            if (Cancelling)
+                throw new CancelledException();
+            try
+            {
+                AddProgressStep(string.Format(Messages.UPDATES_WIZARD_RUNNING_PRECHECK, suppPack, host.Name()));
+
+                PatchPrecheckCheck check = new PatchPrecheckCheck(host, update);
+                var problems = check.RunAllChecks();
+                if (problems != null && problems.Count > 0)
+                {
+                    if (problems[0] is PatchAlreadyApplied)
+                    {
+                        log.InfoFormat("The update {0} is already applied on {1}. Ignore it.", suppPack, host.Name());
+                        ReplaceProgressStep(string.Format(Messages.UPDATES_WIZARD_SKIPPING_UPDATE, suppPack, host.Name()));
+                        alreadyApplied = true;
+                    }
+                    else
+                        throw new Exception(string.Format("{0}: {1}. {2}", host, problems[0].Title,
+                            problems[0].Description));
+                }
+            }
+            catch (Exception ex)
+            {
+                log.Error(string.Format("Precheck failed on host {0}", host.Name()), ex);
+                throw;
+            }
+        }
+
+        private void ApplySuppPack(IXenConnection connection, Session session, string suppPack, Pool_update update)
+        {
+            try
+            {
+                AddProgressStep(string.Format(Messages.UPDATES_WIZARD_APPLYING_UPDATE, suppPack, host.Name()));
+
+                var task = Pool_update.async_apply(session, update.opaque_ref, host.opaque_ref);
+                PollTaskForResultAndDestroy(connection, ref session, task);
+            }
+            catch (Failure f)
+            {
+                if (f.ErrorDescription.Count > 1 && f.ErrorDescription[0] == Failure.UPDATE_ALREADY_APPLIED)
+                {
+                    log.InfoFormat("The update {0} is already applied on {1}. Ignoring this error.", suppPack,
+                        host.Name());
+                    ReplaceProgressStep(string.Format(Messages.UPDATES_WIZARD_SKIPPING_UPDATE, suppPack,
+                        host.Name()));
+                }
+                else
+                    throw;
+            }
+        }
+
+        private void AfterApplySuppPack(Pool_update update)
+        {
+            var afterApplyGuidanceLists = update.after_apply_guidance;
+            if (afterApplyGuidanceLists != null && afterApplyGuidanceLists.Count > 0)
+            {
+                foreach (update_after_apply_guidance afterApplyGuidance in afterApplyGuidanceLists)
+                {
+                    var planAction = GetAfterApplyGuidancePlanAction(host, afterApplyGuidance);
+                    delayedPlanActions.Add(planAction);

```

Comment:
This doesn't take into account the fact that the update might be live-patchable and a reboot might not be needed